torchao/prototype/intx/Intx.py

```diff
@@ -0,0 +1,255 @@
+    return _implements(IntxTensor, aten_ops_or_torch_fn)
+
+def get_layout_tensor_constructor(layout_type_class: type(LayoutType)):
+    return _get_layout_tensor_constructor(AffineQuantizedTensor, layout_type_class)
+
+@dataclass(frozen=True)
+class IntxLayoutType(LayoutType):
+    bit_size: int
+    pack_dim: int = -1
+    
+    def post_process(self, input: torch.Tensor) -> torch.Tensor:
+        from torchao.prototype.intx import to_intx
+        return to_intx(input, self.bit_size, self.pack_dim)
+
+class IntxTensor(torch.Tensor):
+    """
+    Splits int data into packed shards based on bit size
+    fields:
+      int4_shard (torch.Tensor): 4 bit packed shard
+      int2_shard (torch.Tensor): 2 bit packed shard
+      int1_shard (torch.Tensor): 1 bit packed shard
+      bit_size (int): element size in bits
+      pack_dim: (int) dimension to pack along
+    """
+    bits_to_shard = {
+        1: ["int1_shard"],
+        2: ["int2_shard"],
+        3: ["int2_shard", "int1_shard"],
+        4: ["int4_shard"],
+        5: ["int4_shard", "int1_shard"],
+        6: ["int4_shard", "int2_shard"],
+        7: ["int4_shard", "int2_shard", "int1_shard"],
+    }
+    def __new__(
+        cls,
+        shards: List[torch.Tensor],
+        packed_shape: List[int],
+        bit_size: int,
+        pack_dim: int = -1,
+    ):
+        kwargs = {"device": shards[0].device}
+        kwargs["device"] = shards[0].device
+        kwargs["layout"] = shards[0].layout
+        kwargs["requires_grad"] = False
+        kwargs["dtype"] = torch.uint8
+        return torch.Tensor._make_wrapper_subclass(cls, packed_shape, **kwargs) 
+
+    def __init__(
+        self,
+        shards: List[torch.Tensor],
+        packed_shape: List[int],
+        bit_size: int,
+        pack_dim: int = -1,
+    ):
+        shards = [shard.to(torch.uint8) for shard in shards]
+        self.shard = shards
+        for i, atrib in enumerate(self.bits_to_shard[bit_size]):
+            setattr(self, atrib, shards[i])
+            
+        self.packed_shape = packed_shape
+        self.bit_size = bit_size    
+        self.pack_dim = pack_dim
+    
+    def get_shards(self):
+        return [getattr(self,i) for i in self.__class__.bits_to_shard[self.bit_size]]
+    
+    def __repr__(self):
+        return f"Int{self.bit_size}Tensor(shape = {self.packed_shape}, data = {unpack(self.get_shards(), self.bit_size, dim = self.pack_dim)})"
+    
+    def __tensor_flatten__(self):
+        
+        return self.__class__.bits_to_shard[self.bit_size], [self.packed_shape, self.bit_size, self.pack_dim]
+    @classmethod
+    def __tensor_unflatten__(
+        cls, tensor_data_dict, tensor_attributes, outer_size, outer_stride
+    ):
+        shards = [i for i in tensor_data_dict.values()]
+        packed_shape, bit_size, pack_dim = tensor_attributes
+        return cls(shards, packed_shape, bit_size, pack_dim)
+
+    @classmethod
+    def __torch_function__(cls, func, types, args=(), kwargs=None):
```

Comment:
also this with: https://github.com/pytorch/ao/blob/de4a1fb3b1f71e2f61b84dfdc96e7d704ff72208/torchao/dtypes/affine_quantized_tensor.py#L303